src/process.c

```diff
@@ -463,59 +466,450 @@ cc_run_hook(struct oci_cfg_hook* hook, const gchar* state,
+			goto child_failed;
+		}
+
+
+		g_debug ("building hypervisor command-line");
 
+		// FIXME: add network config bits to following functions:
+		//
+		// - cc_oci_container_state()
+		// - oci_state()
+		// - cc_oci_update_options()
+
+		ret = cc_oci_vm_args_get (config, &args);
+		if (! (ret && args)) {
+			goto child_failed;
+		}
+
+		// FIXME: add netcfg to state file
+		ret = cc_oci_state_file_create (config, timestamp);
+		if (! ret) {
+			g_critical ("failed to recreate state file");
+			goto out;
+		}
+
+		g_debug ("running command:");
+		for (p = args; p && *p; p++) {
+			g_debug ("arg: '%s'", *p);
+		}
+
+		if (! cc_oci_setup_child (config)) {
+			goto child_failed;
+		}
+
+		if (execvp (args[0], args) < 0) {
+			g_critical ("failed to exec child %s: %s",
+					args[0],
+					strerror (errno));
+			abort ();
+		}
+	}
+
+	/* parent */
+
+	g_debug ("child pid is %u", (unsigned)pid);
+
+	close (net_cfg_pipe[0]);
+	net_cfg_pipe[0] = -1;
+
+	close (child_err_pipe[1]);
+	child_err_pipe[1] = -1;
+
+	/* create state file before hooks run */
+	ret = cc_oci_state_file_create (config, timestamp);
 	if (! ret) {
-		g_critical ("failed to spawn child: %s", err->message);
-		g_error_free (err);
+		g_critical ("failed to create state file");
+		goto out;
+	}
+
+	/* If a hook returns a non-zero exit code, then an error
+	 * including the exit code and the stderr is returned to
+	 * the caller and the container is torn down.
+	 */
+	ret = cc_run_hooks (config->oci.hooks.prestart,
+			config->state.state_file_path,
+			true);
+	if (! ret) {
+		g_critical ("failed to run prestart hooks");
+	}
+
+	netcfgv = cc_oci_vm_netcfg_to_strv (config);
+	if (! netcfgv) {
+		g_critical ("failed to obtain network config");
+		ret = false;
 		goto out;
+
 	}
 
-	pid = config->state.workload_pid;
+	if (! cc_oci_network_create(config)) {
+		g_critical ("failed to create network");
+		goto out;
+	}
+
+	/* flatten config into a bytestream to send to the child */
+	final = g_strjoinv ("\n", netcfgv);
+	if (! final) {
+		ret = false;
+		goto out;
+	}
+
+	len = strlen (final);
+
+	g_strfreev (netcfgv);
+
+	g_debug ("sending netcfg (%lu bytes) to child",
+			(unsigned long int)len);
+
+	/* Send netcfg to child.
+	 *
+	 * First, send the length of the network
+	 * config data that will follow.
+	 */
+	bytes = write (net_cfg_pipe[1], &len, sizeof (len));
+	if (bytes < 0) {
+		g_critical ("failed to send netcfg length"
+				" to child: %s",
+				strerror (errno));
+		goto out;
+	}
+
+	/* now, send the network config data */
+	bytes = write (net_cfg_pipe[1], final, len);
+	if (bytes < 0) {
+		g_critical ("failed to send netcfg"
+				" to child: %s",
+				strerror (errno));
+		goto out;
+	}
+
+	g_debug ("checking child setup (blocking)");
+
+	/* block reading child error state */
+	bytes = read (child_err_pipe[0],
+			buffer,
+			sizeof (buffer));
+	if (bytes > 0) {
+		g_critical ("child setup failed");
+		ret = false;
+		goto out;
+	}
+
```

Comment:
@mcastelino Same for this block: Having it as a separate routine that takes a `netcfgv`, pushes it through a pipe and wait for the child error state would make the code more readable imho.